src/client/sky.cpp

```diff
@@ -892,19 +831,57 @@ void Sky::setStarCount(u16 star_count, bool force_update)
 	// Allow force updating star count at game init.
 	if (m_star_params.count != star_count || force_update) {
 		m_star_params.count = star_count;
-		m_stars.clear();
-		// Rebuild the stars surrounding the camera
-		for (u16 i = 0; i < star_count; i++) {
-			v3f star = v3f(
-				myrand_range(-10000, 10000),
-				myrand_range(-10000, 10000),
-				myrand_range(-10000, 10000)
-			);
-
-			star.normalize();
-			m_stars.emplace_back(star);
-		}
+		m_seed = (u64)myrand() << 32 | myrand();
+		updateStars();
+	}
+}
+
+void Sky::updateStars() {
+	m_stars.reset(new scene::SMeshBuffer());
+	// Stupid IrrLicht doesn’t allow non-indexed rendering, and indexed quad
+	// rendering is slow due to lack of hardware support. So as indices are
+	// 16-bit and there are 4 vertices per star... the limit is 2^16/4 = 0x4000.
+	// That should be well enough actually.
+	if (m_star_params.count > 0x4000) {
+		warningstream << "Requested " << m_star_params.count << " stars but " << 0x4000 << " is the max\n";
+		m_star_params.count = 0x4000;
+	}
+	m_stars->Vertices.reallocate(4 * m_star_params.count);
+	m_stars->Indices.reallocate(6 * m_star_params.count);
+
+	video::SColor fallback_color = m_star_params.starcolor; // used on GLES 2 “without shaders”
```

Comment:
does "without shaders" here mean that it doesn't use _Minetest's_ shader for stars?

Reply:
This means that that value is only used on GLES 2 when Minetest shaders are disabled and IrrLicht uses its own shaders as GLES 2 can’t work without these. But IrrLicht shaders are broken (now in the sense they don’t implement many features fixed-function OpenGL implements, some of which are marked as TODO and others are silently omitted).